### ISSUE TITLE
FIDO2 in Android app is not working

### ISSUE THREAD

**Issue by numeratorjik** [bitwarden/android#1594]
### Steps To Reproduce

Version 2.13.0 of Android app. Installed on Google Pixel 5 running current Android 12 beta. The app now prompts me for YubiKey login but it doesn't work. First I see this:

![Screenshot_20211018-085344](https://user-images.githubusercontent.com/70389082/137735358-d0ae41b1-6fc3-42da-9b86-e221a2d886a4.png)

I tap "Authenticate WebAuthn" and I see this:

![Screenshot_20211018-085354](https://user-images.githubusercontent.com/70389082/137735450-1d53dca3-32f3-4ebe-b2bb-39ddf8f21e37.png)

I tap "GET STARTED" and I see this:

![Screenshot_20211018-085359](https://user-images.githubusercontent.com/70389082/137735512-68be3971-1295-4da3-b626-08d7f66a458a.png)

When I hold my key up to the back of the phone, the app switches back to the screen shown in the first screenshot above and this page opens in my browser:

![Point Blur_Oct182021_085528](https://user-images.githubusercontent.com/70389082/137735621-50628209-e867-4210-bbaf-a7f9a78eea67.jpg)

Other apps on my phone that use WebAuthn work just fine with my YubiKey. I am able to authenticate to Bitwarden with my YubiKey in my browser on the desktop.

### Expected Result

I expect the app to authenticate successfully with my YubiKey.

### Actual Result

It's not working.

### Screenshots or Videos

See above.

### Operating System

Android

### Operating System Version

Android 12 beta

### Device

Pixel 5

### Build Version

2.13.0

### Beta

- [ ] Using a pre-release version of the application.

**Comment by numeratorjik** [comment]
Please reopen this issue until you have figured out how to make WebAuthn work on Android for Yubikey NEO users.

**Comment by mportune-bw** [comment]
@numeratorjik I agree it's not ideal, but I was unable to find a way to prevent Android's default NDEF scan behavior during my time working on this feature.  Can you provide some examples of apps that are able to do it?  I'm happy to look into it if it's possible.

**Comment by zaneselvans** [comment]
I'm having what seems like the same issue on a Pixel 5a with a Yubikey NEO. I see now that the ... in the upper right hand corner of the screen offers other 2FA options, including using a Yubikey NEO, rather than getting dumped directly to WebAuthn. These other 2FA options should be much more prominent in the flow somehow -- let folks choose which 2FA method to use explicitly. It took me an hour or more of messing around with this before I figured out there were other options.

**Comment by DonRohan** [comment]
> @mportune-bw in response to https://community.bitwarden.com/t/u2f-support-over-nfc/611/53?u=miked I am not sure my problem is the same, I have a fido u2f key (non yubikey) that generally work with webauthn everywhere (and has NFC) I am going through the same flow, getting the OS screen (Choose how to use your security key) and after I authenticate, focus goes back to the app and I see a generic error dialog saying
> 
> > An error has occurred
> 
> I would be happy to record the flow, but Bitwarden has screen capturing protection and can't find a way to disable it
> 
> Might be related [bitwarden/clients#2803](https://github.com/bitwarden/clients/issues/2803)

I'm having the same Issue with a Yubikey 5 NFC and a Solokey Type-C. Webauthn doesnt work with the Bitwarden-App. Even after disabling OTP.

**Comment by improving-rigmarole** [comment]
I'm having the exact same issue on Android 12, impossible to use a yubikey via FIDO2 (OTP is disabled) :
- via NFC : gives the `An error has occurred` popup
- via USB : the OS screen asks for authorization to use USB with the yubikey, asks to press the button, but the led blinks really fast for like 200ms and after this, nothing, and the `press the button` screen stays forever

Any update on this ?

**Comment by kevinjbeattie** [comment]
I opened a bug regarding WebView on Android here: https://issuetracker.google.com/issues/249758200

**Comment by k3a** [comment]
It also didn't work for me with Solo NFC key. Additionally there was no option to select a different 2FA like Authenticator. Such option was available while logging in to the Bitwarden web. I had to disable Solo NFC key in account settings to be able to log in from the Android app.

**Comment by BrendanxP** [comment]
I have the exact same issue and I already unchecked OTP. Bought 5 YubiKey's all with NFC, some with USB-C, but not one works properly on Android.
USB makes the light blink for 200ms and it stops thereafter. No matter if I spam the button instantly or wait a little bit, the key is not picked up. Via NFC it tells me that it succeeded, but then nothing happens. And if i click all windows away the login will give an error. Tried clearing everything from the key and reenabling the FIDO2 auth multiple times. I have no idea what is going wrong here.

**Comment by RZR7332** [comment]
> I've got the same problem as @improving-rigmarole. The led blinks quickly, as if it's failing a handshake or something, then I get this error after a number of seconds. Using default chrome as the webview
> 
> ![Screenshot_20230201-085233.png](https://user-images.githubusercontent.com/161913/216062197-7e1cc74f-8e41-44ec-937c-1e224d6de78c.png)

Exact same problem I am facing as well.

**Comment by mderazon** [comment]
Please see this screencast

https://user-images.githubusercontent.com/717076/217385177-a9256b72-9d67-4313-a1f1-3b7824e71749.mp4

Cannot sign in. Unfortunately, Bitwarden has screen capture protection on by default so the flow starts in the video in black screen from the app, where it goes to the browser and back to the app with failure

**Comment by Jademalo** [comment]
I need to add myself to the voices having problems here, but I think the issue is much bigger than just Bitwarden. Any FIDO2 supporting application is giving me trouble.

To hopefully shed a bit of light on the situation here, after a lot of research and testing it seems that Android's implementation of WebAuthn does not correctly implement CTAP2, and only really implements CTAP1 (U2F).

If the device tries to authenticate with Anrdoid using the FIDO2 CTAP2 protocol, it will fail. 
Due to this, Android flat out cannot handle any situation with Discoverable credentials (formerly called Resident credentials), and [it cannot request a hardware set PIN](https://support.yubico.com/hc/en-us/articles/360016615020-Operating-system-and-web-browser-support-for-FIDO2-and-U2F). In addition, [Android WebView also does not support WebAuthn](https://issuetracker.google.com/issues/249758200?pli=1), and you need to make sure your default browser supports it.

Since Bitwarden's implementation of FIDO2 does not request a pin and can only be used as a 2FA key with a username and password, it theoretically should work on Android. However, **it will only work if the `FIDO U2F` interface is enabled in the YubiKey Manager**. I assume this is because a 2FA only request that does not require a PIN can fall back to CTAP1/U2F and correctly authenticate. [This reddit post](https://old.reddit.com/r/GooglePixel/comments/12td3l4/can_a_pixel_read_a_yubikey/) was what tipped me off to the reasoning.

With FIDO U2F disabled, both Firefox and Chrome give errors, though different. 

Chrome gives the error;
> **An error has occurred.**
> Please make sure your default browser supports WebAuthn and try again.
> 
> NotReadableError: An Unknown error occured while talking to the credential manager.

Firefox gives the error;
> **An error has occurred.**
> Please make sure your default browser supports WebAuthn and try again.
> 
> NotReadableError: The operation failed for an unknown transient reason.


To summarise;
Android does not implement CTAP2, and only implements CTAP1. 
In order to resolve the issue, for both USB and NFC you need to make sure at least FIDO U2F is enabled. FIDO2 does not need to be enabled, but it doesn't seem to affect things if it is.

**Comment by phil-w** [comment]
Same NotAllowedError from the Bitwarden app. 

I'd add that if I use my same default browser (Brave) to access the Bitwarden website itself from Android 13, the FIDO2 2FA also fails in the same way. => I can't use Bitwarden with 2fa via Android, although it works fine from Windows.

**Comment by AlmAck** [comment]
Exact same problem, on android (Pixel5) was working just fine, I had to change the KDF iteration and after that I can't login anymore with the NFC key on my phone. I tried to used also firefox, switch network, reinstall... still nothing. On desktop linux works fine.

**Comment by phil-w** [comment]
As per @AlmAck my keys were marked "migrated" as I'd followed Bitwarden's request to change my "KDF iteration" also. That fix worked for my Android 10 tablet, on which I can now login via browser (Brave, shields up). So at least I can get a password that way.

However... the App still fails as before on my Android 13 phone, and using the same Brave browser also fails there, shields up or down. The symptom is the same - it hangs after the NFC read and if you "back" a few times, you can get back to the web page with the errors in red boxes.

Ah well, so removing then re-adding the keys (a) clears the "migrated" marking, and (b) fixes my Android 10 browser at least. So some progress... that's **at least one thing needs to be added** to the "increase your KDF Iteration" instructions, which is that once you've done it, you need to remove and re-add all FIDO2 keys or they're "migrated", which means "broken".

**Comment by mderazon** [comment]
Bitwarden TWO-STEP LOGIN FIDO2 WebAuthn screens mentiones this warning:
> WARNING
> Due to platform limitations, WebAuthn cannot be used on all Bitwarden applications. You should set up another two-step login provider so that you can access your account when WebAuthn cannot be used. Supported platforms:
> Web vault and browser extensions on a desktop/laptop with a WebAuthn supported browser (Chrome, Opera, Vivaldi, or Firefox with FIDO U2F turned on).

Does that mean that the feature is not supported in mobile at all ?

**Comment by fliespl** [comment]
I just stumbled across this one and failed after 30 minutes to use yubikey 5 nfc. Always reverted to try again screen.

**Comment by callit** [comment]
I was having this problem all day today. I tried a bunch of things, and some combination of the below finally worked. Here's what I did: 
1. Log out of Bitwarden Android app
2. Clear Bitwarden app data
3. Set Chrome as your default browser app for the time being
**4. Before you proceed, I recommend setting up a standard 6-digit OTP authenticator before proceeding in case something goes awry.  You can turn it back off when done if you're satisfied**.
5. Disable NFC OTP as suggested in Bitwarden's troubleshooting document

I believe one of these steps was the actual solution:

7. I had both YubiKey OTP and WebAuthn enabled - Disable YubiKey OTP completely from your two step settings if you have it enabled
8. WebAuthn - My YubiKey NFC had a "Migrated from FIDO" note.  Remove that and re-add your hardware token.
9. Open Bitwarden android app and authenticate with your master password, you'll be prompted for your authenticator - the time it worked for me, instead of authenticating right away, I clicked the menu in the upper right corner and selected "Open In Browser" - the page re-opened in Chrome, and I authenticated with NFC there. I held it until the screen said "you're all set".  It sent me back to the app which was still waiting for the token, so I hit the authenticate button again, and held the NFC to the back of the phone again.

After all of that, it finally let me through. My suspicion is that it was some combination of the YubiKey OTP being enabled and/or having the "Migrated from FIDO" token.

Hopefully this helps someone.

**Comment by vvolkgang** [comment]
Issue migrated to https://github.com/bitwarden/mobile/issues/1594

**Comment by mderazon** [comment]
Downloaded the new beta (native Android app) tried to authenticate with FIDO security key and still not working
I can make a short screencast unfortunately Bitwarden has screen record protection so it will appear blank. 
Would be good to have an option to disable this security feature tucked in somewhere in advanced settings so we can record flows / bugs etc

**Comment by vvolkgang** [comment]
@mderazon can you open a new ticket please? 

Were you able to disable the screen recording protection?

I'll circle back about sharing the video in private.

**Comment by vvolkgang** [comment]
@mderazon do you have an email I can reach out?

**Comment by mderazon** [comment]
@vvolkgang sent you a DM on the community forum
